cpp/src_prims/sparse/distance/csr_spmv.cuh

```diff
@@ -0,0 +1,354 @@
+   * @param row
+   * @param indptrA
+   * @param indicesA
+   * @param dataA
+   */
+  __device__ inline void load_a(value_idx row, value_idx *indptrA,
+                                value_idx *indicesA, value_t *dataA) {
+    if (tid == 0) {
+      offsets_a[0] = indptrA[row];
+      offsets_a[1] = indptrA[row + 1];
+    }
+    __syncthreads();
+
+    start_offset_a = offsets_a[0];
+    stop_offset_a = offsets_a[1];
+
+    shared_size = stop_offset_a - start_offset_a;
+
+    // Coalesce reads of row from matrix A into shared memory
+    for (int i = tid; i < shared_size; i += blockDim.x) {
+      shared_cols[i] = indicesA[start_offset_a + i];
+      shared_vals[i] = dataA[start_offset_a + i];
+    }
+
+    __syncthreads();
+
+    row_a = row;
+  }
+
+  /**
+   * Prepare index & offsets for looping through rows of B
+   * @param start_row
+   * @param indptrB
+   */
+  __device__ inline void load_b(value_idx start_row, value_idx *indptrB) {
+    done = false;
+    shared_idx = 0;
+    cur_sum = 0.0;
+
+    start_row_b = start_row;
+    stop_row_b = min(start_row_b + tpb, start_row_b + (n - start_row_b));
+
+    n_rows = (stop_row_b - start_row_b);
+
+    for (int i = tid; i < n_rows; i += tpb) {
+      row_b = start_row_b + tid;
+      start_offset_b = indptrB[start_row_b + i];
+      row_count = indptrB[start_row_b + i + 1] - start_offset_b;
+      local_idx = start_offset_b;
+      local_idx_stop = start_offset_b + row_count;
+    }
+  }
+
+  /**
+   * Perform single single column intersection/union for A & B
+   * based on the row of A mapped to shared memory and the row
+   * of B mapped to current thread.
+   * @param reduce_func
+   * @param accum_func
+   */
+  __device__ inline void step(reduce_f reduce_func, accum_f accum_func) {
+    if (tid < n_rows) {
+      bool local_idx_in_bounds = local_idx < local_idx_stop && row_count > 0;
+
+      value_idx l = local_idx_in_bounds ? chunk_cols[local_idx] : -1;
+      value_t lv = local_idx_in_bounds ? chunk_vals[local_idx] : 0.0;
+
+      bool shared_idx_in_bounds = shared_idx < shared_size;
+
+      value_idx r = shared_idx_in_bounds ? shared_cols[shared_idx] : -1;
+      value_t rv = shared_idx_in_bounds ? shared_vals[shared_idx] : 0.0;
+
+      bool run_l = ((l <= r && l != -1) || (l != -1 && r == -1));
+      local_idx += 1 * run_l;
+      value_t left_side = lv * run_l;
+
+      bool run_r = ((r <= l && r != -1) || (l == -1 && r != -1));
+      shared_idx += 1 * run_r;
+      value_t right_side = rv * run_r;
+
+      // Apply semiring "sum" & "product" functions locally
+      cur_sum = accum_func(cur_sum, reduce_func(left_side, right_side));
+
+      // finished when all items in chunk have been
+      // processed
+      done = l == -1 && r == -1;
+
+    } else {
+      done = true;
+    }
+  }
+
+  __device__ inline bool isdone() { return done; }
+
+  __device__ inline void write(value_t *out) {
+    for (int i = tid; i < n_rows; i += blockDim.x) {
```

Comment:
In practice the for loop has a single iteration because `n_rows <= blockDim.x`, right? Could we just replace this with an `if (tid < n_rows)`?